iio/iio_ad9361/iio_ad9361.c

```diff
@@ -0,0 +1,2583 @@
+		ret = ad9361_get_rx_rssi (ad9361_phy, channel->ch_num, &rssi);
+		return ret < 0 ? ret : sprintf(buf, "%"PRIu32".%02"PRIu32" dB",
+					       rssi.symbol / rssi.multiplier, rssi.symbol % rssi.multiplier);
+	}
+}
+
+/**
+ * get_hardwaregain_available().
+ * @device:	Physical instance of a iio_axi_adc device.
+ * @buf:	Where value is stored.
+ * @len:	Maximum length of value to be stored in buf.
+ * @channel:	Channel properties.
+ * Return: Length of chars written in buf, or negative value on failure.
+ */
+static ssize_t get_hardwaregain_available(void *device, char *buf, size_t len,
+		const struct iio_ch_info *channel)
+{
+	struct ad9361_rf_phy *ad9361_phy = (struct ad9361_rf_phy *)device;
+
+	if (channel->ch_out) {
+		return (ssize_t) snprintf(buf, len, "[%"PRIi16", %"PRIi16", %"PRIi16"]", 0, 250, 89750);
+	} else {
+		return (ssize_t) snprintf(buf, len, "[%"PRIi32", %"PRIi16", %"PRIi32"]",
+					  ad9361_phy->rx_gain[ad9361_phy->current_table].starting_gain_db,
+					  1,
+					  ad9361_phy->rx_gain[ad9361_phy->current_table].max_gain_db);
+	}
+}
+
+/**
+ * get_sampling_frequency_available().
+ * @device:	Physical instance of a iio_axi_adc device.
+ * @buf:	Where value is stored.
+ * @len:	Maximum length of value to be stored in buf.
+ * @channel:	Channel properties.
+ * Return: Length of chars written in buf, or negative value on failure.
+ */
+static ssize_t get_sampling_frequency_available(void *device, char *buf,
+		size_t len,
+		const struct iio_ch_info *channel)
+{
+	struct ad9361_rf_phy *ad9361_phy = (struct ad9361_rf_phy *)device;
+	int32_t int_dec;
+	uint32_t max;
+
+	if (ad9361_phy->pdata->port_ctrl.pp_conf[2] & LVDS_MODE) {
+		max = 61440000U;
+	} else {
+		max = 61440000U / (ad9361_phy->pdata->rx2tx2 ? 2 : 1);
+	}
+
+	if (channel->ch_out) {
+		if (ad9361_phy->bypass_tx_fir) {
+			int_dec = 1;
+		} else {
+			int_dec = ad9361_phy->tx_fir_int;
+		}
+
+	} else {
+		if (ad9361_phy->bypass_rx_fir) {
+			int_dec = 1;
+		} else {
+			int_dec = ad9361_phy->rx_fir_dec;
+		}
+	}
+
+	return (ssize_t) snprintf(buf, len, "[%"PRIu32" %"PRIi16" %"PRIu32"]",
+				  MIN_ADC_CLK / (12 * int_dec), 1, max);
+}
+
+/**
+ * get_rf_port_select_available().
+ * @device:	Physical instance of a iio_axi_adc device.
+ * @buf:	Where value is stored.
+ * @len:	Maximum length of value to be stored in buf.
+ * @channel:	Channel properties.
+ * Return: Length of chars written in buf, or negative value on failure.
+ */
+static ssize_t get_rf_port_select_available(void *device, char *buf, size_t len,
+		const struct iio_ch_info *channel)
+{
+	if(channel->ch_out) {
+		return (ssize_t) sprintf(buf, "%s %s",
+					 ad9361_rf_tx_port[0],
+					 ad9361_rf_tx_port[1]);
+	} else {
+		ssize_t bytes_no = 0;
+		for(int16_t i = 0; i < sizeof(ad9361_rf_rx_port) / sizeof(ad9361_rf_rx_port[0]);
+		    i++) {
+			if(i > 0 ) {
+				bytes_no += sprintf(buf + bytes_no, " ");
+			}
+			bytes_no += sprintf(buf + bytes_no, "%s", ad9361_rf_rx_port[i]);
+			if(bytes_no < 0) {
```

Comment:
If `sprintf` starts to malfunction after a few iterations `bytes_no` could be big enough to absorb the error which can corrupt the `buf` string.